{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>registry_id</th>\n",
       "      <th>name</th>\n",
       "      <th>ecosystem</th>\n",
       "      <th>description</th>\n",
       "      <th>homepage</th>\n",
       "      <th>licenses</th>\n",
       "      <th>repository_url</th>\n",
       "      <th>normalized_licenses</th>\n",
       "      <th>versions_count</th>\n",
       "      <th>...</th>\n",
       "      <th>rankings</th>\n",
       "      <th>namespace</th>\n",
       "      <th>advisories</th>\n",
       "      <th>maintainers_count</th>\n",
       "      <th>first_release_published_at</th>\n",
       "      <th>keywords</th>\n",
       "      <th>docker_dependents_count</th>\n",
       "      <th>docker_downloads_count</th>\n",
       "      <th>issue_metadata</th>\n",
       "      <th>latest_version</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1408</th>\n",
       "      <td>8420236</td>\n",
       "      <td>58</td>\n",
       "      <td>Organism.dplyr</td>\n",
       "      <td>bioconductor</td>\n",
       "      <td>dplyr-based Access to Bioconductor Annotation ...</td>\n",
       "      <td></td>\n",
       "      <td>Artistic-2.0</td>\n",
       "      <td></td>\n",
       "      <td>[Artistic-2.0]</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>{'downloads': 19.847856154910097, 'dependent_r...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-10-12 14:50:27.601000+00:00</td>\n",
       "      <td>[Annotation, GenomeAnnotation, Sequencing, Sof...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>{'id': 89095484, 'package_id': 8420236, 'numbe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>920</th>\n",
       "      <td>8420313</td>\n",
       "      <td>58</td>\n",
       "      <td>PING</td>\n",
       "      <td>bioconductor</td>\n",
       "      <td>Probabilistic inference for Nucleosome Positio...</td>\n",
       "      <td></td>\n",
       "      <td>Artistic-2.0</td>\n",
       "      <td></td>\n",
       "      <td>[Artistic-2.0]</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>{'downloads': 31.800262812089358, 'dependent_r...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-10-12 14:52:04.943000+00:00</td>\n",
       "      <td>[Clustering, Sequencing, Software, Statistical...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>{'id': 89095567, 'package_id': 8420313, 'numbe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>795</th>\n",
       "      <td>8419992</td>\n",
       "      <td>58</td>\n",
       "      <td>msa</td>\n",
       "      <td>bioconductor</td>\n",
       "      <td>Multiple Sequence Alignment</td>\n",
       "      <td>http://www.bioinf.jku.at/software/msa/</td>\n",
       "      <td>GPL (&gt;= 2)</td>\n",
       "      <td>http://www.bioinf.jku.at/software/msa/</td>\n",
       "      <td>[CNRI-Python-GPL-Compatible]</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>{'downloads': 8.498168498168498, 'dependent_re...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>2023-10-12 14:46:53.252000+00:00</td>\n",
       "      <td>[Alignment, MultipleComparison, MultipleSequen...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>{'id': 89095108, 'package_id': 8419992, 'numbe...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           id  registry_id            name     ecosystem  \\\n",
       "1408  8420236           58  Organism.dplyr  bioconductor   \n",
       "920   8420313           58            PING  bioconductor   \n",
       "795   8419992           58             msa  bioconductor   \n",
       "\n",
       "                                            description  \\\n",
       "1408  dplyr-based Access to Bioconductor Annotation ...   \n",
       "920   Probabilistic inference for Nucleosome Positio...   \n",
       "795                         Multiple Sequence Alignment   \n",
       "\n",
       "                                     homepage      licenses  \\\n",
       "1408                                           Artistic-2.0   \n",
       "920                                            Artistic-2.0   \n",
       "795   http://www.bioinf.jku.at/software/msa/     GPL (>= 2)   \n",
       "\n",
       "                               repository_url           normalized_licenses  \\\n",
       "1408                                                         [Artistic-2.0]   \n",
       "920                                                          [Artistic-2.0]   \n",
       "795   http://www.bioinf.jku.at/software/msa/   [CNRI-Python-GPL-Compatible]   \n",
       "\n",
       "      versions_count  ...                                           rankings  \\\n",
       "1408               1  ...  {'downloads': 19.847856154910097, 'dependent_r...   \n",
       "920                1  ...  {'downloads': 31.800262812089358, 'dependent_r...   \n",
       "795                1  ...  {'downloads': 8.498168498168498, 'dependent_re...   \n",
       "\n",
       "     namespace advisories maintainers_count       first_release_published_at  \\\n",
       "1408       NaN         []                 1 2023-10-12 14:50:27.601000+00:00   \n",
       "920        NaN         []                 1 2023-10-12 14:52:04.943000+00:00   \n",
       "795        NaN         []                 1 2023-10-12 14:46:53.252000+00:00   \n",
       "\n",
       "                                               keywords  \\\n",
       "1408  [Annotation, GenomeAnnotation, Sequencing, Sof...   \n",
       "920   [Clustering, Sequencing, Software, Statistical...   \n",
       "795   [Alignment, MultipleComparison, MultipleSequen...   \n",
       "\n",
       "     docker_dependents_count docker_downloads_count issue_metadata  \\\n",
       "1408                     NaN                    NaN           None   \n",
       "920                      NaN                    NaN           None   \n",
       "795                      NaN                    NaN           None   \n",
       "\n",
       "                                         latest_version  \n",
       "1408  {'id': 89095484, 'package_id': 8420236, 'numbe...  \n",
       "920   {'id': 89095567, 'package_id': 8420313, 'numbe...  \n",
       "795   {'id': 89095108, 'package_id': 8419992, 'numbe...  \n",
       "\n",
       "[3 rows x 35 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import networkx as nx\n",
    "from tqdm import tqdm\n",
    "\n",
    "df = pd.read_json(\"../data/bioconductor.ndjson\", lines=True)\n",
    "df[\"latest_release_published_at\"] = df[\"latest_release_published_at\"].astype(str)\n",
    "df.sample(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['id', 'registry_id', 'name', 'ecosystem', 'description', 'homepage',\n",
       "       'licenses', 'repository_url', 'normalized_licenses', 'versions_count',\n",
       "       'latest_release_published_at', 'latest_release_number',\n",
       "       'keywords_array', 'language', 'status', 'last_synced_at', 'created_at',\n",
       "       'updated_at', 'metadata', 'repo_metadata', 'repo_metadata_updated_at',\n",
       "       'dependent_packages_count', 'downloads', 'downloads_period',\n",
       "       'dependent_repos_count', 'rankings', 'namespace', 'advisories',\n",
       "       'maintainers_count', 'first_release_published_at', 'keywords',\n",
       "       'docker_dependents_count', 'docker_downloads_count', 'issue_metadata',\n",
       "       'latest_version'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': 89092443,\n",
       " 'package_id': 8418609,\n",
       " 'number': '1.68.0',\n",
       " 'published_at': '2023-10-12T14:10:28.218Z',\n",
       " 'licenses': None,\n",
       " 'integrity': None,\n",
       " 'status': None,\n",
       " 'created_at': '2023-10-12T14:10:28.218Z',\n",
       " 'updated_at': '2023-10-12T14:10:28.218Z',\n",
       " 'metadata': {},\n",
       " 'dependencies': [{'id': 975828933,\n",
       "   'package_id': None,\n",
       "   'version_id': 89092443,\n",
       "   'package_name': 'graphics',\n",
       "   'ecosystem': 'bioconductor',\n",
       "   'kind': 'imports',\n",
       "   'optional': False,\n",
       "   'requirements': '*'},\n",
       "  {'id': 975828934,\n",
       "   'package_id': None,\n",
       "   'version_id': 89092443,\n",
       "   'package_name': 'grDevices',\n",
       "   'ecosystem': 'bioconductor',\n",
       "   'kind': 'imports',\n",
       "   'optional': False,\n",
       "   'requirements': '*'},\n",
       "  {'id': 975828935,\n",
       "   'package_id': None,\n",
       "   'version_id': 89092443,\n",
       "   'package_name': 'methods',\n",
       "   'ecosystem': 'bioconductor',\n",
       "   'kind': 'imports',\n",
       "   'optional': False,\n",
       "   'requirements': '*'},\n",
       "  {'id': 975828937,\n",
       "   'package_id': None,\n",
       "   'version_id': 89092443,\n",
       "   'package_name': 'stats',\n",
       "   'ecosystem': 'bioconductor',\n",
       "   'kind': 'imports',\n",
       "   'optional': False,\n",
       "   'requirements': '*'},\n",
       "  {'id': 975828938,\n",
       "   'package_id': None,\n",
       "   'version_id': 89092443,\n",
       "   'package_name': 'tcltk',\n",
       "   'ecosystem': 'bioconductor',\n",
       "   'kind': 'imports',\n",
       "   'optional': False,\n",
       "   'requirements': '*'},\n",
       "  {'id': 975828939,\n",
       "   'package_id': None,\n",
       "   'version_id': 89092443,\n",
       "   'package_name': 'utils',\n",
       "   'ecosystem': 'bioconductor',\n",
       "   'kind': 'imports',\n",
       "   'optional': False,\n",
       "   'requirements': '*'},\n",
       "  {'id': 975828940,\n",
       "   'package_id': 8419763,\n",
       "   'version_id': 89092443,\n",
       "   'package_name': 'limma',\n",
       "   'ecosystem': 'bioconductor',\n",
       "   'kind': 'suggests',\n",
       "   'optional': False,\n",
       "   'requirements': '*'},\n",
       "  {'id': 975828941,\n",
       "   'package_id': 8419782,\n",
       "   'version_id': 89092443,\n",
       "   'package_name': 'LPE',\n",
       "   'ecosystem': 'bioconductor',\n",
       "   'kind': 'suggests',\n",
       "   'optional': False,\n",
       "   'requirements': '*'},\n",
       "  {'id': 975828936,\n",
       "   'package_id': 8420112,\n",
       "   'version_id': 89092443,\n",
       "   'package_name': 'multtest',\n",
       "   'ecosystem': 'bioconductor',\n",
       "   'kind': 'imports',\n",
       "   'optional': False,\n",
       "   'requirements': '*'},\n",
       "  {'id': 975828932,\n",
       "   'package_id': 8418836,\n",
       "   'version_id': 89092443,\n",
       "   'package_name': 'Biobase',\n",
       "   'ecosystem': 'bioconductor',\n",
       "   'kind': 'imports',\n",
       "   'optional': False,\n",
       "   'requirements': '*'}]}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.latest_version.iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 1420/1420 [00:10<00:00, 133.06it/s]\n"
     ]
    }
   ],
   "source": [
    "processed_nodes = set()\n",
    "graph = nx.Graph()\n",
    "for _, row in tqdm(df.iterrows(), total=len(df)):\n",
    "    # Get current node name\n",
    "    node_name = row[\"name\"]\n",
    "\n",
    "    # Fast handle none\n",
    "    if node_name is None:\n",
    "        continue\n",
    "\n",
    "    # Create node if package not in graph\n",
    "    if node_name not in processed_nodes:\n",
    "        graph.add_node(\n",
    "            node_name,\n",
    "            last_release=row.latest_release_published_at,\n",
    "            downloads=row.downloads,\n",
    "        )\n",
    "        processed_nodes.add(node_name)\n",
    "\n",
    "    # For each dependency in the \"latest_version\" column,\n",
    "    # check if the dependency is already in the graph, if not, add it.\n",
    "    # Then add an edge between the package and the dependency.\n",
    "    dependency_names = []\n",
    "    for dep in row.latest_version[\"dependencies\"]:\n",
    "        # Fast handle none\n",
    "        if dep is None or dep[\"package_name\"] is None:\n",
    "            continue\n",
    "    \n",
    "        # Get dep name\n",
    "        dep_name = dep[\"package_name\"]\n",
    "\n",
    "        # Handle not in graph\n",
    "        if dep_name not in processed_nodes:\n",
    "            dep_details = df[df[\"name\"] == dep_name]\n",
    "            if len(dep_details) == 0:\n",
    "                continue\n",
    "            \n",
    "            else:\n",
    "                dep_details = dep_details.iloc[0]\n",
    "                last_published = dep_details.latest_release_published_at\n",
    "                graph.add_node(\n",
    "                    dep_name,\n",
    "                    last_release=last_published,\n",
    "                    downloads=dep_details.downloads,\n",
    "                )\n",
    "                processed_nodes.add(dep_name)\n",
    "\n",
    "        # Add edge\n",
    "        graph.add_edge(node_name, dep_name)\n",
    "\n",
    "nx.write_gexf(graph, \"bioconductor.gexf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "czi-hack",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
